src/core/queriesObserver.ts

```diff
@@ -64,20 +64,24 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map(options => {
+    return queries.map((options, index) => {
       const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions).getOptimisticResult(
+      return this.getObserver(defaultedOptions, index).getOptimisticResult(
         defaultedOptions
       )
     })
   }
 
-  private getObserver(options: QueryObserverOptions): QueryObserver {
+  private getObserver(
+    options: QueryObserverOptions,
+    index: number
+  ): QueryObserver {
     const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-    return (
-      this.observersMap[defaultedOptions.queryHash!] ||
-      new QueryObserver(this.client, defaultedOptions)
-    )
+    let currentObserver = this.observersMap[defaultedOptions.queryHash!]
+    if (!currentObserver && defaultedOptions.keepPreviousData) {
+      currentObserver = this.observers[index]
```

Comment:
please let me know if this is a bad idea. Generally, I was thinking that if we have keepPreviousData on, we would like to keep the data of the query that lies at the index. This likely won't work if the lengths are completely different, and we might "keep" wrong data then. But by just going by the `queryHash` like before, we're never able to make `keepPreviousData` work because we'll get a new Observer every time as the hash changes ...

Reply:
One way to limit this would be to only do the index access if the length hasn’t changed. That would make keepPreviousData only work on equal lengths though and we should likely document that